libs/parser/parser.c

```diff
@@ -0,0 +1,409 @@
+		free(result->fields);
+		hmap_free(result->headersInfosMap);
+		free(result->headerInfos);
+		free(result);
+		return NULL;
+	}
+
+	if (parser_common.refs == 0) {
+		if (parser_compileRegex() != 0) {
+			hmap_free(result->fieldsMap);
+			free(result->fields);
+			hmap_free(result->headersInfosMap);
+			free(result->headerInfos);
+			free(result);
+			return NULL;
+		}
+	}
+
+	parser_common.refs++;
+	return result;
+}
+
+
+static int parser_compileRegex()
+{
+	if (regcomp(&(parser_common.commentRegex), COMMENT_PATTERN, REG_EXTENDED) != 0) {
+		return -1;
+	}
+	if (regcomp(&(parser_common.fieldRegex), FIELD_PATTERN, REG_EXTENDED) != 0) {
+		regfree(&(parser_common.commentRegex));
+		return -1;
+	}
+	if (regcomp(&(parser_common.headerRegex), HEADER_PATTERN, REG_EXTENDED) != 0) {
+		regfree(&(parser_common.commentRegex));
+		regfree(&(parser_common.fieldRegex));
+		return -1;
+	}
+
+	return 0;
+}
+
+
+void parser_free(parser_t *p)
+{
+	if (p == NULL) {
+		return;
+	}
+
+	hmap_free(p->headersInfosMap);
+	free(p->headerInfos);
+
+	hmap_free(p->fieldsMap);
+	free(p->fields);
+
+	free(p);
+
+	parser_common.refs--;
+	if (parser_common.refs == 0) {
+		regfree(&(parser_common.commentRegex));
+		regfree(&(parser_common.fieldRegex));
+		regfree(&(parser_common.headerRegex));
+	}
+}
+
+
+int parser_headerAdd(parser_t *p, const char *headerName, int (*converter)(hmap_t *))
+{
+	if (p == NULL || headerName == NULL || converter == NULL) {
+		return -1;
+	}
+
+	/* Checking if there is enough memory for a new header */
+	if (p->headersInfosMap->capacity == p->headersInfosMap->size) {
+		return -1;
+	}
+
+	if (strlen(headerName) > MAX_HEADER_LEN) {
+		return -1;
+	}
+
+	p->headerInfos[p->headersInfosMap->size].converter = converter;
+	strcpy(p->headerInfos[p->headersInfosMap->size].headerName, headerName);
+
+	if (hmap_insert(p->headersInfosMap,
+			p->headerInfos[p->headersInfosMap->size].headerName,
+			&(p->headerInfos[p->headersInfosMap->size])) == -1) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+/* Returns a type of line from file */
+enum parser_lineType parser_lineTypeGet(char *line)
+{
+	enum parser_lineType result = lineType_InvalidLine;
+
+	if (regexec(&(parser_common.fieldRegex), line, 0, NULL, 0) == 0) {
+		result = lineType_Field;
+	}
+	else if (regexec(&(parser_common.commentRegex), line, 0, NULL, 0) == 0) {
+		result = lineType_Comment;
+	}
+	else if (regexec(&(parser_common.headerRegex), line, 0, NULL, 0) == 0) {
+		result = lineType_Header;
+	}
+
+	return result;
+}
+
+
+/* Returns next header name */
+static int parser_headerGet(char *line, char *result)
+{
+	regmatch_t regmatch[HEADER_SUBEXPRESSIONS];
+	int headerLen;
+
+	if (regexec(&(parser_common.headerRegex), line, HEADER_SUBEXPRESSIONS, regmatch, 0) == 0) {
+		headerLen = regmatch[1].rm_eo - regmatch[1].rm_so;
+
+		if (headerLen > MAX_HEADER_LEN) {
+			return -1;
+		}
+
+		strncpy(result, &line[regmatch[1].rm_so], headerLen);
+		result[headerLen] = '\0';
+
+		return 0;
+	}
+
+	return -1;
```

Comment:
In our code style, we try to limit the amount of returns in the code.

```suggestion
        int err = -1;
	if (regexec(&(parser_common.headerRegex), line, HEADER_SUBEXPRESSIONS, regmatch, 0) == 0) {
		headerLen = regmatch[1].rm_eo - regmatch[1].rm_so;
		if (headerLen < MAX_HEADER_LEN) {
		        strncpy(result, &line[regmatch[1].rm_so], headerLen);
		        result[headerLen] = '\0';
		        err = 0;
		}
	}

	return err;
```